feature(authority): write block import messages on a shared queue

arminsabouri opened the pull request:
PoA consensus will consume block import messages and notify the beacon engine of new blocks. As well as update its own cache and prune the mempool

oobiora committed:
start PoA scaffolding

oobiora committed:
WIP: define epoch manager responsabilities

oobiora committed:
WIP: need to define signer in epoch_manager

oobiora committed:
WIP: start AuthorityTask

oobiora committed:
WIP: main task

arminsabouri committed:
fix: use botanix_lib to deserialize extra data header in authority/constants

arminsabouri committed:
fix: instantiate epoch manager in authority consensus builder

arminsabouri committed:
refactor: add genesis extradata header

arminsabouri committed:
fix: calculate block sighash after block header is completed

arminsabouri committed:
fix: skip withdrawals root consensus check if chain id is botanix_testnet

arminsabouri committed:
feature(authority): validate signer limit util

arminsabouri committed:
consensus: load consensus cache with headers till start of POA epoch

arminsabouri committed:
refactor(authority_consensus): refactor block production task to be a async fn

arminsabouri committed:
feature(authority): Use block import to share blocks between authority peers

arminsabouri committed:
feature(authority): validate incoming blocks

arminsabouri committed:
refactor: include recent bitcoin block hash in botanix extra data header field

arminsabouri committed:
fix: validate header on authority block import

arminsabouri committed:
feature(authority): update storage consensus on new blocks

arminsabouri committed:
refactor: remove unsed `PerBlockConstants` from authority consensus crate

arminsabouri committed:
refactor: update calls to `extraDataHeader` encode/deocde to use the new interface

arminsabouri committed:
feature(authority): write block import messages on a shared queue

PoA consensus will consume block import messages and notify the beacon
engine of new blocks. As well as update its own cache and prune the
mempool

arminsabouri committed:
fix(net/import): forward new headers from peers

arminsabouri committed:
refactor(consensus): improve logging in common consensus validation

arminsabouri committed:
feature(authority): execute txs and call fcu

stevenroose commented on crates/consensus/authority/src/task.rs:
```diff
@@ -105,6 +115,79 @@ where
+                                        storage.insert_new_block(block.header, body);
+                                        let mining_pool = self.pool.clone();
+                                        // Lastly remove confirmed txs from the mempool
+                                        mining_pool.remove_transactions(
```
nit: why do you clone the pool Arc instead of just do `self.pool.remove_transactions()`?

stevenroose commented on crates/consensus/authority/src/task.rs:
```diff
@@ -105,6 +115,79 @@ where
+                            payload: block.clone().seal_slow().into(),
+                            cancun_fields: None,
+                            tx,
+                        });
```
I think you should panic here on the error for `send`. It means that the receiving thread has stopped (probably paniced, but might be gracefully shutting down). I'm not 100% sure but otherwise the `await` down there might block this thread forever.

stevenroose replied:
Was this ignored on purpose?

arminsabouri replied:
Not purposfully. Sorry I missed your comments on this PR

arminsabouri replied:
Not purposfully. Sorry I missed your comments on this PR

arminsabouri replied:
Yeah thats a good point. If the engine API is down or cleaning up there isnt much we can do in the Consensus layer anyways

arminsabouri replied:
Yeah thats a good point. If the engine API is down or cleaning up there isnt much we can do in the Consensus layer anyways

stevenroose commented on crates/consensus/authority/src/task.rs:
```diff
@@ -105,6 +115,79 @@ where
             let mut storage = self.storage.write().await;
 
+            match self.block_import_rx.try_recv() {
+                Ok(new_block) => {
```
Just a style thing, but I think it's easier to follow the code if the error cases from this method are not like 100 lines back down. Also, leaving the `match` asap saves you 2 levels of indentation. I would write this like
```rust
            let new_block = match self.block_import_rx.try_recv() {
                Ok(b) => b,
                Err(TryRecvError::Empty) => {
                    debug!(target: "consensus::authority", "No new blocks from peers");
                    break;
                },
                Err(TryRecvError::Disconnected) => {
                    error!(target: "consensus::authority", "Block import channel disconnected");
                    return ();
                },
            };
```            
And then the rest of the code can just be below that instead of inside the `Ok() => {    }` block.

stevenroose replied:
I think it helps that while reading the `try_recv` you immediatelly see how errors are handled instead of having to scroll down and find them. Also in this care you have 3 following error handling things following each other so you'd also have to track indentation to see which belongs to which:
![image](https://github.com/botanix-labs/Macbeth/assets/853468/a1b73853-5d54-4dff-bcce-360df46cc668)

stevenroose replied:
Did this here https://github.com/botanix-labs/Macbeth/pull/72

stevenroose commented on crates/consensus/authority/src/task.rs:
```diff
@@ -105,6 +115,79 @@ where
+                            tx,
+                        });
+
+                        match rx.await.unwrap() {
```
Same thing here the error handling can be earlier and can save you 2 levels of indentation.

arminsabouri committed:
refactor(authority): do not construct headers with non-zero difficulty

arminsabouri committed:
fix: compiling issues after upstream rebases

arminsabouri committed:
fix: remove extra_data_header size check in payload engine

This check can be re-introduced when botanix protocol has a better cap
on extra_data_header size.
Commit also removes unused function commented out that was moved
else where
